safekeeper/src/bin/safekeeper.rs

```diff
@@ -116,11 +121,20 @@ fn main() -> anyhow::Result<()> {
         .get_one::<String>("auth-validation-public-key-path")
         .map(PathBuf::from);
 
+    if let Some(log_format) = arg_matches.get_one::<String>("log-format") {
+        conf.log_format = LogFormat::from_str(log_format).with_context(|| {
+            format!(
+                "Unrecognized log format. Please specify one of: [{}]",
+                get_log_formats()
```

Comment:
```suggestion
                "Unrecognized log format. Please specify one of: {:?}",
                LogFormat::VARIANTS
```

I think this would do the same trick, and allow removing `get_log_formats` fn. The `{:?}` ends up calling through trait `std::fmt::Debug` to render the value, which for slices is `[{:?}, ...]`, which would probably be `["plain", "json"]` in this case.

Same applies for the pageserver usage of `get_log_formats`.

Looking more, now because the `LogFormat::from_str` impl is generated by a proc macro, we need to repeat the anyhow message for it in different callsites. Might be worthwhile to consider a `LogFormat::from_config(&str) -> anyhow::Result<()>` which would be the one place to do this.

Reply:
Wanted to avoid adding a dependencies, since neither pageserver or safekeeper had `strum`, which seems required to have access to `VARIANTS`.

> Might be worthwhile to consider a LogFormat::from_config(&str) -> anyhow::Result<()> which would be the one place to do this.
Thank you, I was looking for a way to customize the error, but couldn't figure out. I'll look in to wrapping it.

Reply:
Adding a dependency for this particular thing is a bit of extra, I personally would had just written it manually since it seems quite unlikely that we'd introduce a third while forgetting to update that list. Then again it is impossible to write a test for the list going out of sync :) Well, I guess one could use a macro to define the enum but ... Maybe it's easier to just have this dependency, which apparently was already in our `Cargo.lock`.

> I'll look in to wrapping it.

I just meant that moving the repeated thing as `LogFormat::from_config(s: &str) -> anyhow::Result<LogFormat>` would do the trick. Using `anyhow`'s context for this is fine.

Reply:
>  I personally would had just written it manually

The more I think about it, the more this feels the right choice, because it's also the least fragile and most trivial thing to change if it ever even needs to be changed. And at the same time having the least possible negative effect :)

Reply:
so KISS it is :)

Reply:
Thank you for explaining how anyhow works, and now I think I understood what you really meant. Should look much better?

Reply:
Looking better! Resolving.